### DISCUSSION TITLE
Using declarations from other Pest files.

### DISCUSSION THREAD

**Discussion by RonaldBunk** [pest-parser/pest#844] [Q&A] [answered]
As the title states, I'd like to be able to use pre-existing declarations in other Pest files, for example:

base.pest
```pest
STRING = _{ "\"" ~ ANY+ "\"" }
```

some-other.pest
```pest
@use "base" // Or implicitly imported when it's found in the same directory.
SCHEMA = { SOI ~ "[" ~ base::STRING ~ "]" ~ EOI }

@use { STRING } from "base" // Or implicitly imported when it's found in Pest files in the same directory.
SCHEMA = { SOI ~ "[" ~ STRING ~ "]" ~ EOI }
```

**Comment by RonaldBunk** [comment]
In additionally, perhaps it would be an option to create namespaces, for example like:
```pest
namespace "base" {
  STRING = _{ "\"" ~ ANY+ "\"" }
}

@use "base"
@use { STRING } from "base"
```

**Comment by tomtau** [comment, marked as answer]
https://pest.rs/book/grammars/grammars.html#load-multiple-grammars

**Reply by RonaldBunk** [reply]
That still doesn't seem to allow the usage of declarations from one file to be used in another, merely having all of it available at code-level.

I mainly asked this because a single file of declarations becomes very large and when using multiple files, I still have to re-declare anything that I want to use in each file individually.

To which I thought that it might be more practical if previous declared definitions can be re-used.

**Reply by tomtau** [reply]
It allows reusing AFAIK. See e.g. https://github.com/pest-parser/pest/blob/master/derive/examples/base.pest which is then used in https://github.com/pest-parser/pest/blob/master/derive/examples/calc.pest 

https://github.com/pest-parser/pest/blob/master/derive/examples/calc.rs

**Reply by RonaldBunk** [reply]
Thank you, I didn't know! 

My editor complained about those symbols missing. I think that's an issue in regard to the plug-in/extension (I use VS Code & IntelliJ IDEA).

**Reply by tomtau** [reply]
No problem, it's a relatively recent feature, so the IDE extensions may not support it yet. I opened an issue on https://github.com/pest-parser/pest-ide-tools/issues/24 for the VS Code extension.
For IntelliJ (https://github.com/pest-parser/intellij-pest), I don't know if @ice1000 has time to fix it to support this feature.
